File: low-level-design/4-structural-design-patterns/2-decorator-pattern.py
# BAD EXAMPLE
# Each combination of pizza requires a new class
class PlainPizza:
    pass

class CheesePizza(PlainPizza):
    pass

class OlivePizza(PlainPizza):
    pass

class StuffedPizza(PlainPizza):
    pass

class CheeseStuffedPizza(CheesePizza):
    pass

class CheeseOlivePizza(CheesePizza):
    pass

class CheeseOliveStuffedPizza(CheeseOlivePizza):
    pass

# GOOD EXAMPLE  
from abc import ABC, abstractmethod

# =========== Component Interface ============
class Pizza(ABC):
    @abstractmethod
    def getDescription(self):
        pass

    @abstractmethod
    def getCost(self):
        pass


# ============= Concrete Components: Base pizza ==============
class PlainPizza(Pizza):
    def getDescription(self):
        return "Plain Pizza"

    def getCost(self):
        return 150.00


class MargheritaPizza(Pizza):
    def getDescription(self):
        return "Margherita Pizza"

    def getCost(self):
        return 200.00


# ======================== Abstract Decorator ===========================
# ====== Implements Pizza and holds a reference to a Pizza object =======
class PizzaDecorator(Pizza):
    def __init__(self, pizza):
        self.pizza = pizza


# ============ Concrete Decorator: Adds Extra Cheese ================
class ExtraCheese(PizzaDecorator):
    def getDescription(self):
        return self.pizza.getDescription() + ", Extra Cheese"

    def getCost(self):
        return self.pizza.getCost() + 40.0


# ============ Concrete Decorator: Adds Olives ================
class Olives(PizzaDecorator):
    def getDescription(self):
        return self.pizza.getDescription() + ", Olives"

    def getCost(self):
        return self.pizza.getCost() + 30.0


# ============ Concrete Decorator: Adds Stuffed Crust Cheese ================
class StuffedCrust(PizzaDecorator):
    def getDescription(self):
        return self.pizza.getDescription() + ", Stuffed Crust"

    def getCost(self):
        return self.pizza.getCost() + 50.0



if __name__ == "__main__":
    # BAD: Using PlainPizza and decorators to create a pizza
    plainPizza = PlainPizza()

    # Pizzas with individual toppings
    cheesePizza = CheesePizza()
    olivePizza = OlivePizza()
    stuffedPizza = StuffedPizza()

    # Combinations of toppings require separate classes
    cheeseStuffedPizza = CheeseStuffedPizza()
    cheeseOlivePizza = CheeseOlivePizza()

    # Further combinations increase complexity exponentially
    cheeseOliveStuffedPizza = CheeseOliveStuffedPizza()

    # GOOD: Using decorators to create a pizza
    # Start with a basic Margherita Pizza
    myPizza = MargheritaPizza()

    # Add Extra Cheese
    myPizza = ExtraCheese(myPizza)

    # Add Olives
    myPizza = Olives(myPizza)

    # Add Stuffed Crust
    myPizza = StuffedCrust(myPizza)

    # Final Description and Cost
    print("Pizza Description:", myPizza.getDescription())
    print("Total Cost: ₹" + str(myPizza.getCost()))

File: low-level-design/4-structural-design-patterns/2-decorator-pattern.java
import java.util.*;

// BAD EXAMPLE
// Each combination of pizza requires a new class
class PlainPizza {}
class CheesePizza extends PlainPizza {}
class OlivePizza extends PlainPizza {}
class StuffedPizza extends PlainPizza {}
class CheeseStuffedPizza extends CheesePizza {}
class CheeseOlivePizza extends CheesePizza {}
class CheeseOliveStuffedPizza extends CheeseOlivePizza {}

// public class Main {
//     public static void main(String[] args) {
//         // Base pizza
//         PlainPizza plainPizza = new PlainPizza();

//         // Pizzas with individual toppings
//         CheesePizza cheesePizza = new CheesePizza();
//         OlivePizza olivePizza = new OlivePizza();
//         StuffedPizza stuffedPizza = new StuffedPizza();

//         // Combinations of toppings require separate classes
//         CheeseStuffedPizza cheeseStuffedPizza = new CheeseStuffedPizza();
//         CheeseOlivePizza cheeseOlivePizza = new CheeseOlivePizza();

//         // Further combinations increase complexity exponentially
//         CheeseOliveStuffedPizza cheeseOliveStuffedPizza = new CheeseOliveStuffedPizza();

//     }
// }


// GOOD EXAMPLE
interface Pizza {
    String getDescription();
    double getCost();
}

class PlainPizzaGood implements Pizza {
    @Override
    public String getDescription() {
        return "Plain Pizza";
    }

    @Override
    public double getCost() {
        return 150.00;
    }
}

class MargheritaPizza implements Pizza {
    @Override
    public String getDescription() {
        return "Margherita Pizza";
    }

    @Override
    public double getCost() {
        return 200.00;
    }
}

// Abstract Decorator
// Implements Pizza and holds a reference to a Pizza object
abstract class PizzaDecorator implements Pizza {
    protected Pizza pizza;

    public PizzaDecorator(Pizza pizza) {
        this.pizza = pizza;
    }
}

// Concrete Decorators
class ExtraCheese extends PizzaDecorator {
    public ExtraCheese(Pizza pizza) {
        super(pizza);
    }

    @Override
    public String getDescription() {
        return pizza.getDescription() + ", Extra Cheese";
    }

    @Override
    public double getCost() {
        return pizza.getCost() + 30.00;
    }
}

class Olives extends PizzaDecorator {
    public Olives(Pizza pizza) {
        super(pizza);
    }

    @Override
    public String getDescription() {
        return pizza.getDescription() + ", Olives";
    }

    @Override
    public double getCost() {
        return pizza.getCost() + 20.00;
    }
}

class StuffedCrust extends PizzaDecorator {
    public StuffedCrust(Pizza pizza) {
        super(pizza);
    }

    @Override
    public String getDescription() {
        return pizza.getDescription() + ", Stuffed Crust";
    }

    @Override
    public double getCost() {
        return pizza.getCost() + 50.0;
    }
}

public class Main {
    public static void main(String[] args) {

        // BAD: Using PlainPizzaGood and decorators to create a pizza
        // Each combination of pizza requires a new class
        // Base pizza
        PlainPizza plainPizza = new PlainPizza();

        // Pizzas with individual toppings
        CheesePizza cheesePizza = new CheesePizza();
        OlivePizza olivePizza = new OlivePizza();
        StuffedPizza stuffedPizza = new StuffedPizza();

        // Combinations of toppings require separate classes
        CheeseStuffedPizza cheeseStuffedPizza = new CheeseStuffedPizza();
        CheeseOlivePizza cheeseOlivePizza = new CheeseOlivePizza();

        // Further combinations increase complexity exponentially
        CheeseOliveStuffedPizza cheeseOliveStuffedPizza = new CheeseOliveStuffedPizza();

        // GOOD: Using decorators to create a pizza
        Pizza pizza = new PlainPizzaGood();
        pizza = new ExtraCheese(pizza);
        pizza = new Olives(pizza);
        pizza = new StuffedCrust(pizza);

        System.out.println(pizza.getDescription());
        System.out.println(pizza.getCost());
    }
}